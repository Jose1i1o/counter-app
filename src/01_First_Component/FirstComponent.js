// import { Fragment } from 'react'; // Fragment is a component and is an unnecessary import

// we have two kind of components: those based on classes and those based in components
// The last are now called functional components, previously SFC (Stainless Functional Components)
// With hooks Functional Components are much easier to maintain and to read

// This is a FC
// const FirstComponent = ( props ) => { // We can use destructuring here
const FirstComponent = ({ greetings = "There is a bird on my window shelf ü¶úü•∞" }) => {

  // console.log(props);
  // const greetings = "Have a great day! ‚ù§Ô∏è"; 

  const person = { 
    name: 'Jose',
    age: 33
  }

  // To use variables we will use {} in the HTML syntax. We can use strings, number, arrays but not booleans. 

  // Return needs the parenthesis if we have a space, breakdown between return and the html or if there is more than one tag. All the code also needs to be contained within one tag/ element to keep the reference easily. For example a <div></div>.

  // However, creating a <div></div> will generate unnecessary elements. This can be solved in different ways. For example, using a <fragment></fragment> which needs to be imported from React, the other way is with a <></>

  // See the html in the console to see the different inserted elements
  return (
    <>
    {/* <h1> { props.greetings } </h1> */}
    <h1> { greetings } </h1>
    {/* <h1> { props.greetings } </h1> This is comming from the props at index.js */}
    {/* <pre> { JSON.stringify(person) } </pre> Text in the tag pre is shown in a fixed font, it¬¥s not proportional, it is exactly shown as it is in the file */}
    <span>My name is { person.name } and I hope you are ok</span>
  </>
  )
}

export default FirstComponent;
