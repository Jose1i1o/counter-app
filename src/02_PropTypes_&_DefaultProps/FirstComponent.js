import PropTypes from 'prop-types';

const FirstComponent = ({ greetings, subtitle="Today is Monday, 23rd of May 2022" }) => {

// using conditional will result in memory loss. To avoid this we will use moduleName
//   if (!greetings || greetings === "") {
//     throw new Error("You need to pass a greetings prop to the component");
// }

  return (
    <>
      <h1> { greetings } </h1>
      <span>{ subtitle }</span>
    </>
  )
}

// This is a validation that will require a prop to be passed or wonÂ´t compile or will throw error
FirstComponent.propTypes = {
  // greetings: PropTypes.string // This is optional
  greetings: PropTypes.string.isRequired // This is required
}

// This is a default prop. They are typically left to empty so that you will know there is a prop to be completed
// When there is one, this will over rule any other prop you pass to it
FirstComponent.defaultProps = {
  subtitle: "My birthday is on the 30th of May"
}

export default FirstComponent;
