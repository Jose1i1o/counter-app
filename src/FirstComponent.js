import React, { Fragment } from 'react';
import PropTypes from 'prop-types';

const FirstComponent = ({ greetings, subtitle }) => {
  return (
    <>
      <h1> { greetings } </h1>
      <p> { subtitle } </p>
    </>
  )
}

  FirstComponent.propTypes = {
    greetings: PropTypes.string.isRequired
  }

  FirstComponent.defaultProps = {
    subtitle: 'I am a subtitle'
  }

export default FirstComponent;