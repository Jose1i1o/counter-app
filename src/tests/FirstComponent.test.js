import React from 'react';
import { render } from '@testing-library/react';
import FirstComponent from "../FirstComponent.js";
// import { shallow } from 'enzyme';
// for enzyme we also need npm i --save-dev enzyme-to-json

describe('Testing my first react Component <FirstComponent />', () => {
  test('It should display the message "Lola is my dog´s name"', () => {

    const greeting = "Lola is my dog´s name";


    const { getByText } = render( <FirstComponent greetings={ greeting }/>);
    // eslint-disable-next-line testing-library/prefer-screen-queries
    expect( getByText(greeting) ).toBeInTheDocument();
  })
})


// use enzyme to test the component

// Test 1
// describe('Testing my first react Component <FirstComponent /> with enzyme', () => {
//   test('it should display <FirstComponent /> correctly', () => {
//     const greeting = "Lola is my dog´s name";
//     const wrapper = shallow(<FirstComponent greetings={ greeting }/>);

//     expect( wrapper).toMatchSnapshot();
//   })
// })

// Test 2
// describe('Testing my first react Component <FirstComponent /> with enzyme', () => {
//   test('it should display <FirstComponent /> correctly', () => {
//   const greeting = 'My dog´s name is Lola';
//   const subheader = 'This is a subtitle';

//   const wrapper = shallow(
//       <FirstComponent greetings={ greeting } subtitle={ subheader }/>
//   );

//   const paragraph = wrapper.find('p').text();
//   expect( paragraph ).toBe( subheader );
//   })
// })
