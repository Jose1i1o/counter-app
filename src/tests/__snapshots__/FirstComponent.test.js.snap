// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing in <FirstComponent /> should show it correctly 1`] = ``
<Fragment>
  <h1>

    My dogÂ´s name is Lola

  </h1>
  <p>

    This is a subtitle

  </p>
</Fragment>
`;